_inc/lib/class-jetpack-tweetstorm-helper.php

```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * Tweetstorm block and API helper.
+ *
+ * @package jetpack
+ */
+
+use Automattic\Jetpack\Connection\Client;
+
+/**
+ * Class Jetpack_Tweetstorm_Helper
+ */
+class Jetpack_Tweetstorm_Helper {
+	/**
+	 * Gather the Tweetstorm.
+	 *
+	 * @param  string $url The tweet URL to gather from.
+	 * @return mixed
+	 */
+	public static function gather( $url ) {
+		$site_id = self::get_site_id();
+		if ( is_wp_error( $site_id ) ) {
+			return $site_id;
+		}
+
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			if ( ! class_exists( 'WPCOM_Gather_Tweetstorm' ) ) {
+				\jetpack_require_lib( 'gather-tweetstorm' );
+			}
+
+			return WPCOM_Gather_Tweetstorm::gather( $url );
+		}
+
+		$response = Client::wpcom_json_api_request_as_blog(
+			sprintf( '/sites/%d/tweetstorm/gather?url=%s', $site_id, rawurlencode( $url ) ),
+			2,
+			array( 'headers' => array( 'content-type' => 'application/json' ) ),
+			null,
+			'wpcom'
+		);
+		if ( is_wp_error( $response ) ) {
+			return $response;
+		}
+
+		$data = json_decode( wp_remote_retrieve_body( $response ) );
+
+		if ( wp_remote_retrieve_response_code( $response ) >= 400 ) {
+			return new WP_Error( $data->code, $data->message, $data->data );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Get the WPCOM or self-hosted site ID.
+	 *
+	 * @return mixed
+	 */
+	public static function get_site_id() {
+		$is_wpcom = ( defined( 'IS_WPCOM' ) && IS_WPCOM );
+		$site_id  = $is_wpcom ? get_current_blog_id() : Jetpack_Options::get_option( 'id' );
+		if ( ! $site_id ) {
+			return new WP_Error(
+				'unavailable_site_id',
+				__( 'Sorry, something is wrong with your Jetpack connection.', 'jetpack' ),
+				403
```

Comment:
This might be a silly question, but is `403` the best status here?

Reply:
I copy/pasted this function from `Jetpack_Instagram_Gallery_Helper`, which probably got it from `WPCOM_REST_API_V2_Endpoint_Mailchimp`, I didn't really think about the error code it's using. 😉

The site ID being unavailable could be caused by a number of things, some of them are appropriate for a `403` status (as the message suggests, if  the Jetpack connection needs fixing, `403` is correct). Alternatively, `502` or a generic `500` response may be more accurate for whatever has caused this error.

I don't think this method the right place to be debugging why there's no valid site id, even though the site has been connected to WPCOM (as is evident by the block loading), so `403` is a reasonable approximation of what could be causing it.

Reply:
I could see cases where it fit and some where it might not have so well. I agree that it is likely the best approximation as you say.

Thanks for the explanation 🙂